structx/build.rs

```diff
@@ -18,124 +18,267 @@ use std::{
     path::PathBuf,
 };
 
-use syn::{
-    FnArg,
-    ItemFn,
-    Macro,
-    Member,
-    Pat,
-    PatMacro,
-    Type,
-    visit::{self, Visit},
-};
+use syn::{Error, Expr, FnArg, GenericArgument, ItemFn, Macro, Member, Meta, Pat, Path, PathArguments, PatStruct, ReturnType, Type, visit::{self, Visit}};
+use syn::parse::{Parse, ParseStream};
+use syn::spanned::Spanned;
+
+/*
+ * Introduced a new type which can be parsed with syn::parse2.
+ * This is necessary because syn version 2 doesn't implement Parse for Pat
+ */
+struct StructX {
+    pat: PatStruct
+}
+impl Parse for StructX {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let pat = Pat::parse_single(input)?;
+        if let Pat::Struct(pat) = pat {
+            Ok(StructX { pat })
+        } else {
+            Err(Error::new(
+                pat.span(),
+                "structx!()'s supported pattern matching is struct only."
+            ))
+        }
+    }
+}
 
 fn wrap_struct_name( struct_name: &str, input: TokenStream ) -> TokenStream {
     let mut ts = Ident::new( struct_name, Span::call_site() ).into_token_stream();
     ts.extend( Some( TokenTree::Group( Group::new( Delimiter::Brace, input ))));
     ts
 }
 
-fn join_fields( fields: impl Iterator<Item=(Ident, Option<Type>)> ) -> (String, Vec<Ident>, Vec<Option<Type>>) {
+fn join_fields( fields: impl Iterator<Item=Ident> ) -> (String, Vec<Ident>) {
     let mut fields = fields.collect::<Vec<_>>();
-    fields.sort_by_key( |field| field.0.clone() );
+    fields.sort_by_key( |field| field.clone() );
     fields
         .into_iter()
         .fold(
-            ("structx".to_owned(), Vec::new(), Vec::new() ),
-            |(mut struct_name, mut field_idents, mut field_types), (ident, ty)| {
+            ("structx".to_owned(), Vec::new()),
+            |(mut struct_name, mut field_idents), ident| {
                 let field_name = ident.to_string();
-                struct_name.push( '_' );
-                struct_name.push_str( &field_name.replace( "_", "__" ));
-                field_idents.push( ident.clone() );
-                field_types.push( ty );
-                (struct_name, field_idents, field_types)
+                struct_name.push('_');
+                struct_name.push_str(&field_name.replace( "_", "__" ));
+                field_idents.push(ident);
+                (struct_name, field_idents)
             }
         )
 }
 
-type StructMap = HashMap<String, (Vec<Ident>, Vec<Option<Type>>)>;
+type StructMap = HashMap<String, Vec<Ident>>;
 
 struct StructxCollector<'a>( &'a mut StructMap );
 
 impl<'a> Visit<'_> for StructxCollector<'a> {
-    fn visit_macro( &mut self, mac: &Macro ) {
```

Comment:
just changed the order of visit_macro and visit_item_fn because intellij recommended it